WEB: Fix maintainers grid not displaying correctly (GH41438)

calvh opened the pull request:
- [x] closes #41438 
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them

Screenshot taken on Google Chrome 90.0.4430.93

![screenshot](https://user-images.githubusercontent.com/18680207/118068760-ef740800-b370-11eb-902d-237da39e2b7f.png)

To reproduce, 
- go to https://pandas.pydata.org/about/team.html
- open developer tools
- go to elements tab
- navigate to <div class="row maintainers">
- edit as HTML
- delete <div class="row maintainers"> and the closing tag at the end

calvh committed:
WEB: Remove row div from maintainers (GH41438)

attack68 commented:
have you tried instead:

```
    <div class="card-group maintainers">
        {% for person in maintainers.people %}
                <div class="card">
                ...
                </div>
        {% endfor %}
    </div>
```

seems to cut out maybe a lot of unnecessary batch grouping?

calvh commented:
Thanks! I tried it but I ended up with very small cards. See:
![card-group_only](https://user-images.githubusercontent.com/18680207/118143750-ad32e100-b3d9-11eb-8e82-2eb087322f43.png)

I can use CSS to set `min-width:10rem;`. See:
![min-width_option](https://user-images.githubusercontent.com/18680207/118149622-b32bc080-b3df-11eb-888d-8c602dbf58c2.png)

Would this be preferable?

MarcoGorelli commented:
> Edit: What is the recommended way to edit PRs?

If I've understood the question correctly: by just adding a new commit (they'll all get merged at the end anyway)

calvh commented:
> > Edit: What is the recommended way to edit PRs?
> 
> If I've understood the question correctly: by just adding a new commit (they'll all get merged at the end anyway)

Thanks. I was wondering if I should amend my local commit and force push but I don't know how it'll affect this PR (or if I should make a new PR). Or do something like squashing commits.

calvh commented:
After further testing, setting min-width alone doesn't work either because if you set the screen size wide enough the last row (Patrick) ends up takes up the whole row . I'll tinker further to see if I can come up with a solution.

datapythonista commented:
Thanks @calvh 

This should be what we want: https://bootstrapcreative.com/pattern/responsive-card-deck-example/

The CI should be now fix if you merge master into your branch.

calvh commented:
I've come up with a solution that keeps the structure as:
```html
<div class="card-group maintainers">
  <div class="card">
  <div class="card">
  ...
```
by adding the following CSS:
```css
@media (min-width: 576px) {
    div.card {
        min-width: 10rem;
        max-width: 10rem;
    }
}
```

See: 
![demo](https://user-images.githubusercontent.com/18680207/118167186-6b636400-b3f4-11eb-82ad-979f56977c95.gif)

If I want to avoid changing the CSS and use classes like in the example given by @datapythonista I could use something like:
```html
<div class="row row-cols-1 row-cols-sm-2 row-cols-md-3 row-cols-lg-4 row-cols-xl-5">
  <div class="col mb-1">
  <div class="card"
  ...
```
but that requires a higher version of bootstrap (e.g., 4.6). Current version is 4.0. 

Should I update bootstrap and follow the example?

datapythonista commented:
If you want to upgrade bootstrap, that surely sounds good. But let's have a separate PR for the upgrade, and another one for this fix. Up to you.

calvh commented:
I think upgrading Bootstrap is a good idea. They just dropped a stable version (5.0) and it will open up new features for future developers to improve the website. In addition, v5 no longer uses jQuery so one less dependency is better in my opinion. I will do some testing on my local machine.

datapythonista commented:
Yes, a new issue can help give it more visibility, in case anyone else has opinions.

calvh committed:
WEB: Fix maintainers grid display issue (GH41438)

Overwrites previous commit
Uses a single card-group container
Uses CSS to fix width issues

datapythonista commented on web/pandas/about/team.md:
```diff
@@ -8,31 +8,27 @@ If you want to support pandas development, you can find information in the [dona
-        </div>
+<div class="card-group maintainers">
+    {% for person in maintainers.people %}
+        {% if person %}
```
If I'm not wrong, this was the `if` used to discriminate the maintainers from the empty loop iterations generated by Jinja's `batch`, so not needed anymore. All the maintainers are people now ;)

calvh replied:
Good catch, thanks. My bad. I'll make the change on the inside as well.

datapythonista commented on web/pandas/static/css/pandas.css:
```diff
@@ -45,6 +45,12 @@ a.navbar-brand img {
 div.card {
     margin: 0 0 .2em .2em !important;
 }
+@media (min-width: 576px) {
```
Out of curiosity. What is this `min-width`?

calvh replied:
This min-width after the "@media" means that the enclosed rules will only apply if the screen width is _at least_ 576 px. 

i.e., if the screen is less than 576 px, then there are no rules applied. See:
![less_than_576](https://user-images.githubusercontent.com/18680207/118185978-2a2a7e80-b40b-11eb-95fa-67cc1d6efd2b.png)

when the screen is more than 576 px wide, then the rules get applied. See:

![more_than_576](https://user-images.githubusercontent.com/18680207/118186010-357daa00-b40b-11eb-9e2f-bbf23c21d5e4.png)

I can also make it so that the rules always apply no matter the width. See:

![always_fixed_width](https://user-images.githubusercontent.com/18680207/118186324-9c02c800-b40b-11eb-8378-8dc210b6205f.png)

I'm not sure which is better so I just followed the behavior of the existing website.

datapythonista reviewed (commented):
Just one thing that can be simplified. But if this is the code used for the animation you shared, looks really great.

calvh committed:
WEB: Remove if-else block in template (GH41438)

datapythonista reviewed (approved):
Thanks @calvh, looks great!

attack68 reviewed (approved):
I think there are many ways you could do this, and this looks good to me as is.

Subjectively, each card is small enough to fit on even a small smartphone display, so personally I would code all the cards to be same size regardless of display device and design each card so that text wrapping etc, was good looking, and then just use the flex-wrap to align everything (as you already have done), which for a smartphone might be 1 or two cards per row, for a tablet 4 or 5, and for a large display display 6 - 10. I'd use the flex-box `justify-content` so that for a single row it aligns to the center. But, still can do any of these as a follow on (or not) and its good enough for me!

calvh commented:
@datapythonista Please remove approval for now there is a conflict with the community/blog page because it also uses the div.card style. I'll try to fix this asap

datapythonista reviewed (changes requested):
Thanks @calvh, let me know when it's ready

calvh committed:
WEB: Add CSS selector to width rule (GH41438)

Add .card-group.maintainers to limit rule to team page

jreback commented:
@datapythonista pls merge when ready

calvh commented:
Just pushed. I fixed it by adding additional selectors to the new rule so it only applies to the team page and not the blog page.

datapythonista reviewed (approved):
Thanks for the fix @calvh, and for double checking the websie. Looks good. Feel free to ping me once the CI is green and I'll merge.

datapythonista commented:
Thanks for the fix @calvh. If you want to have a look at the production website in couple of hours (after it's updated), and see if everything looks as expected, that would be great.

calvh commented:
Your welcome I am happy to do so later. Thanks for the support!

datapythonista commented:
Looks like the grid if now displayed correctly, and Patrick's card is shown as the others. The rest of the website looks as before. Thanks for the fix @calvh
